### ISSUE TITLE
Build script to concat them all

### ISSUE THREAD

**Issue by fabianmoronzirfas** [ExtendScript/extendscript-es5-shim#2]
This needs some kind of build script that concats all files in subdirectories.

**Comment by EugenTepin** [comment]
I suggest webpack for that purpose.

**Comment by fabianmoronzirfas** [comment]
I never started using webpack. Doesn't it wrap all the files into some magical function? Will we run into errors with Extendscript on that? I try to avaid tools like grunt, gulp, webpack, broccoli and what not if possible. 

If it doesn't mean to have some config overhead, lets go for it but would a simple `cat ./**/*.js > bundle.js` do the trick? (Just a question. I'm not against it but wnat to keep this as light weight as possible with little or no dependencies)

**Comment by fabianmoronzirfas** [comment]
@vogelino wrote this concat.js script https://github.com/fabianmoronzirfas/extendscript.prototypes/blob/master/npm/concat.js

**Comment by EugenTepin** [comment]
If you wait about 9-10 hours i will demonstrate workflow using webpack in [this](https://github.com/EugenTepin/ps-scripting-es5shim) repo.

**Comment by fabianmoronzirfas** [comment]
Cool. I'll wait.

**Comment by EugenTepin** [comment]
1.  Clone [this repo](https://github.com/EugenTepin/ps-scripting-es5shim) in your local repo
2.  Comand line: `npm install` . 
3. See **webpack.config.js**.
4. See **index.js** (it is quite simple)
5. Run `webpack` in comand line (there is also watch mode: webpack _--watch_) or use ready bundles
6. **Open test.bundle.js in ExtendScript toolkit**
7. Run it.
8. See console output.

Later i will write my thoughts about production usage.

**Comment by fabianmoronzirfas** [comment]
Hi, thanks for the example. Looks and works good (in a quick test in ESTK and InDesign).

I'm still not to convinced about all the webpack bootstrap function stuff around the actual functions. I would love to hear the opinion from other members as well. If the majority is okay with it is fine for me.  

Another question is how to go on with the development. You said

> But i prefer to keep my repo and commit into it.

How can we then integrate that into one combined effort? Any thoughts on that?

**Comment by EugenTepin** [comment]
Hi! 

> I'm still not to convinced about all the webpack bootstrap function stuff around the actual functions.

I sopouse that you will have to use wraper functions anyway, just because it is a good practise not to populate/expose variables to global scope.

**Comment by EugenTepin** [comment]
> I would love to hear the opinion from other members as well. If the majority is okay with it is fine for me.

Absolutely agree with you. More opinions is better, than only two.

As for my words about repo. They only means that i will develop it anyway (with its name and infrastructure: npm, webpack etc). And ofcourse i will help with this repo, but first we need to clarify some questions:

- What structure will be in this repo?
- What tools we will use for development? 

I mean: Which bundlers we will use? Or we will not use any of them or self made. With what tools we are going to test it? And so on and so forth.

**Comment by fabianmoronzirfas** [comment]
> I suppose that you will have to use wrapper functions anyway, just because it is a good practise not to populate/expose variables to global scope.  

But wouldn't that create a opinionated package? I - for example - have often a build process that wraps the final script into a wrapper function. This would mean I have two wrapper functions when we wrap the whole thing.  

If I would write this just for myself I would go for a plain file where all the functions are in the global scope and then wrap them together with the rest of my script.

**Comment by vogelino** [comment]
I think the great advantage of bundling in the case of @ExtendScript/extendscriptr is to be able to combine multiple, smaller files into one executable script, avoiding the one-file-chaos. In the case of [es5-shim](https://github.com/ExtendScript/extendscript-es5-shim), I wonder if this is really necessary and also expect concatenation to be enougth.

**Comment by EugenTepin** [comment]
Sorry, i was offline for a while.

> But wouldn't that create a opinionated package? I - for example - have often a build process that wraps the final script into a wrapper function. This would mean I have two wrapper functions when we wrap the whole thing. 

 I have an idea about it. **What if the structure of this package will be developed environment agnostic (as much as possible)?** For example, it will have the following structure:

>Global object to patch (it is a folder. Example **Array**)
>>method/property of this object that is shimmed (it is a file with .js extension. Example: **forEach**.js)
>> ...


>Another Global object to patch (it is a folder. Example **Function**)
>>....

This gives us the following:

- you may include in your build/app only shims you needed 
- small files are easier in development and support
- you can develop in parallel

_Then, when you install this package, you build it according to your own environment or dev workflow._ Everybody happy!
What do you say about that?

**Comment by fabianmoronzirfas** [comment]
Hi @EugenTepin sounds lovely.  
I still would suggest having a simple build script that concats them all. The structure you suggest is like this right?

      .
      └── root
          ├── dist
          ├── Array
          ├── Function
          └── Object

**Comment by EugenTepin** [comment]
Hi, @fabianmoronzirfas . I prefer  something like that:
```
root
    |-distr
    |-lib
         |-Array
         |-Function
         |-Object
```
But I am OK with yours variant too.

**Comment by fabianmoronzirfas** [comment]
Maybe we provide different ways of bundling the prototypes?  I'm not sure if this is a solution. I would love to hear from other members but we can't force them to give their 50cents

**Comment by EugenTepin** [comment]
> Maybe we provide different ways of bundling the prototypes? I'm not sure if this is a solution. 

I didn't get the point, what do you mean by that?

> I would love to hear from other members but we can't force them to give their 50cents

Maybe you should **_mention_** them? Or it would be not correct? I think we can`t get their opinion because they simply  do not know about this conversation.

**Comment by fabianmoronzirfas** [comment]
> I didn't get the point, what do you mean by that?

Maybe we provide a webpack build script and a plain concat build script and so on. But this would also mean to keep them up to date which is extra work.  

> About the mention. Yeah probably we should. Hi @ExtendScript/core members. We could need some input on on what build process would be the most useful for you. Just reply with a 👍  on the following comments. Thanks.

**Comment by fabianmoronzirfas** [comment]
webpack with wrapper functions

**Comment by fabianmoronzirfas** [comment]
- simple concat with all shims in global scope

**Comment by fabianmoronzirfas** [comment]
something completely different (please explain)

**Comment by vogelino** [comment]
I am still not sure if there is a way to bundle them with webpack without running into errors. Did you try this? I think to remember of a bug with object.defineProperty missing but needed by webpack or something similar. To know this would help to make a decision. Unfortunately, I have no time to try this right now.

**Comment by fabianmoronzirfas** [comment]
It worked with this example @EugenTepin wrote https://github.com/EugenTepin/ps-scripting-es5shim/blob/master/webpack.config.js

**Comment by vogelino** [comment]
Ok, I get it. Still, didn't the shims need to be available globally? What would be the way (or a possible way) of including the shims in a project or, let's say, @ExtendScript/extendscriptr for example.  Would we need to `import 'extendscript-es5shim';` in every file? Only once in the entry files? How would the API look like in this case, as bundling prevents the polyfills to be available in the global scope.

**Comment by EugenTepin** [comment]
Hi, @vogelino. I try to clarify your questions. 
1) Any code that patches something, **must be executed first.** 
2) As for global scope, **this patch applied to the global objects** it is in the global scope already :) . See the example : 
```
Object.hello = "Anybody home?";

(function(){
// wrapper #1

    (function(){
        // we need to go deeper :)

            (function(){
                // the limbo :)
                    Object.hello = "Hello world";
                })()
        
        })();
    
    })();

alert(Object.hello);
```
3) As for package usage, @fabianmoronzirfas correct me if i mistaken, i see it right so:

- Package installs with builded version and as seperate files, so you can grab all shims or just only that you need.
- If you want to maintain the script (commit something) you need to install devDependences, wich are something that builds shims into one bundle. And this is the subject of our discussion with @fabianmoronzirfas

**Comment by fabianmoronzirfas** [comment]
@vogelino : @EugenTepin is right. We can't import the shim in an ExtendScript via import or require. That's why I also would go for a simple concat for the shim and leave it to the user how to use them in their ExtendScript.   
Might be:  

    //@include './node_modules/extendscript-es5-shim/dist/extendscript-es5-shim.bundle.js'

**Comment by fabianmoronzirfas** [comment]
Well. At least we got a reaction from @wridgers that counts 2:1 concat vs webpack  
https://github.com/ExtendScript/extendscript-es5-shim/issues/2#issuecomment-251109360  
Let's wait a little more and then start working on the structure.

**Comment by vogelino** [comment]
Sorry I am still confused: Even if the bundled webpack version cannot be imported via require it still can be bundled and included like you showed us right?

    //@include './node_modules/extendscript-es5-shim/dist/extendscript-es5-shim.webpack-bundle.js'

**Comment by fabianmoronzirfas** [comment]
The bundled version can be included via the extendscript `//@include` or `#include` statement. It worked also with the webpack version @EugenTepin provided in his repo

**Comment by vogelino** [comment]
All right now I understand how this we work technically on both side (I'm quite slow I know 😝 ). I'm still not sure about the use of webpack if this does not really enable something in particular or do I still miss something?

**Comment by EugenTepin** [comment]
Well, by now, i will leave you for a while. I need to work on the polyfill by it self. And, if i will have any free time, i will create demo repo wich will express my concept of developing scripts for Photoshop.
I think that until actual scripts will not be ready, you can use dummy-scripts (plain files maybe) to write the concat script to build them into bundle, keeping in mind [this repo structure](https://github.com/ExtendScript/extendscript-es5-shim/issues/2#issuecomment-251083906).

**Comment by EugenTepin** [comment]
Polyfill is ready, but we need to test it. I created an [issue](https://github.com/EugenTepin/ps-scripting-es5shim/issues/2) in my repo for basic test, anybody who wants to test are welcome. As for this repo, i am ready to transfer code here from my repo at any time (before or after test) according to structure described in [this](https://github.com/ExtendScript/extendscript-es5-shim/issues/2#issuecomment-251083906) comment by @fabianmoronzirfas.

**Comment by fabianmoronzirfas** [comment]
Hi guys. I'm currently in Quito. Will be back next week and then it will take me some time to get on track again with all that work that got stuck. :-) So don't be mad if you don't hear back from me in the next two weeks.

**Comment by fabianmoronzirfas** [comment]
Finally. I added the concat script.  closed by 90492d2
